src/Clash/Cores/Rgmii.hs

```diff
@@ -0,0 +1,216 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+
+{- |
+Module      : Clash.Cores.Rgmii
+Description : Functions and types to connect an RGMII PHY to a packet stream interface.
+
+To keep this module generic, users will have to provide their own "primitive" functions:
+
+    1. delay functions to set to the proper amount of delay (which can be different for RX and TX);
+    2. iddr function to turn a single DDR (Double Data Rate) signal into 2 non-DDR signals;
+    3. oddr function to turn two non-DDR signals into a single DDR signal.
+
+Note that Clash models a DDR signal as being twice as fast, thus both facilitating
+and requiring type-level separation between the two "clock domains".
+-}
+module Clash.Cores.Rgmii (
+  RgmiiRxChannel (..),
+  RgmiiTxChannel (..),
+  rgmiiReceiver,
+  rgmiiSender,
+  unsafeRgmiiRxC,
+  rgmiiTxC,
+) where
+
+import Clash.Prelude
+
+import Protocols
+import Protocols.PacketStream
+
+import Data.Maybe (isJust)
+
+-- | RX channel from the RGMII PHY
+data RgmiiRxChannel dom domDDR = RgmiiRxChannel
+  { rgmiiRxClk :: "rx_clk" ::: Clock dom
+  , rgmiiRxCtl :: "rx_ctl" ::: Signal domDDR Bit
+  , rgmiiRxData :: "rx_data" ::: Signal domDDR (BitVector 4)
+  }
+
+instance Protocol (RgmiiRxChannel dom domDDR) where
+  type Fwd (RgmiiRxChannel dom domDDR) = RgmiiRxChannel dom domDDR
+  type Bwd (RgmiiRxChannel dom domDDR) = Signal dom ()
+
+-- | TX channel to the RGMII PHY
+data RgmiiTxChannel domDDR = RgmiiTxChannel
+  { rgmiiTxClk :: "tx_clk" ::: Signal domDDR Bit
+  , rgmiiTxCtl :: "tx_ctl" ::: Signal domDDR Bit
+  , rgmiiTxData :: "tx_data" ::: Signal domDDR (BitVector 4)
+  }
+
+instance Protocol (RgmiiTxChannel domDDR) where
+  type Fwd (RgmiiTxChannel domDDR) = RgmiiTxChannel domDDR
+  type Bwd (RgmiiTxChannel domDDR) = Signal domDDR ()
+
+-- | RGMII receiver.
+rgmiiReceiver ::
+  forall dom domDDR.
+  (DomainPeriod dom ~ (*) 2 (DomainPeriod domDDR)) =>
+  (KnownDomain dom) =>
+  -- | RX channel from the RGMII PHY
+  RgmiiRxChannel dom domDDR ->
+  -- | RX delay function
+  (forall a. Signal domDDR a -> Signal domDDR a) ->
+  -- | iddr function
+  ( forall a.
+    (NFDataX a, BitPack a) =>
+    Clock dom ->
+    Reset dom ->
+    Signal domDDR a ->
+    Signal dom (a, a)
+  ) ->
```

Comment:
I note this function doesn't match the signature of `Clash.Explicit.DDR.ddrIn`. To use it, one would need to do something like:

```haskell
... ddrInComp ...
 where
    ddrInComp = \clk rst -> ddrIn clk rst enableGen (deepErrorX "Undefined initial value")
```

I'd suggest adding `Enable` and reset value to the `iddr` function. Same thing for `oddr`.